jbi/services/bugzilla.py

```diff
@@ -0,0 +1,128 @@
+        resp.raise_for_status()
+        parsed = resp.json()
+        if parsed.get("error"):
+            raise BugzillaClientError(parsed["message"])
+        return parsed
+
+    @property
+    def logged_in(self) -> bool:
+        """Verify the API key validity."""
+        # https://bugzilla.readthedocs.io/en/latest/api/core/v1/user.html#who-am-i
+        resp = self._call("GET", f"{self.base_url}/rest/whoami")
+        return "id" in resp
+
+    def get_bug(self, bugid) -> BugzillaBug:
+        """Retrieve details about the specified bug id."""
+        # https://bugzilla.readthedocs.io/en/latest/api/core/v1/bug.html#rest-single-bug
+        url = f"{self.base_url}/rest/bug/{bugid}"
+        bug_info = self._call("GET", url)
+        parsed = BugzillaApiResponse.parse_obj(bug_info)
+        if not parsed.bugs:
+            raise BugzillaClientError(
+                f"Unexpected response content from 'GET {url}' (no 'bugs' field)"
+            )
+        bug = parsed.bugs[0]
+        # If comment is private, then fetch it from server
+        if bug.comment and bug.comment.is_private:
+            comment_list = self.get_comments(bugid)
+            matching_comments = [c for c in comment_list if c.id == bug.comment.id]
+            # If no matching entry is found, set `bug.comment` to `None`.
+            found = matching_comments[0] if matching_comments else None
+            bug = bug.copy(update={"comment": found})
+        return bug
+
+    def get_comments(self, bugid) -> list[BugzillaComment]:
+        """Retrieve the list of comments of the specified bug id."""
+        # https://bugzilla.readthedocs.io/en/latest/api/core/v1/comment.html#rest-comments
+        url = f"{self.base_url}/rest/bug/{bugid}/comment"
+        comments_info = self._call("GET", url)
+        comments = comments_info.get("bugs", {}).get(str(bugid), {}).get("comments")
+        if comments is None:
+            raise BugzillaClientError(
+                f"Unexpected response content from 'GET {url}' (no 'bugs' field)"
+            )
+        return parse_obj_as(list[BugzillaComment], comments)
+
+    def update_bug(self, bugid, **fields) -> BugzillaBug:
+        """Update the specified fields of the specified bug."""
+        # https://bugzilla.readthedocs.io/en/latest/api/core/v1/bug.html#rest-update-bug
+        url = f"{self.base_url}/rest/bug/{bugid}"
+        updated_info = self._call("PUT", url, json=fields)
+        parsed = BugzillaApiResponse.parse_obj(updated_info)
+        if not parsed.bugs:
+            raise BugzillaClientError(
+                f"Unexpected response content from 'PUT {url}' (no 'bugs' field)"
+            )
+        return parsed.bugs[0]
+
+
+instrumented_methods = (
+    "getbug",
+    "get_comments",
+    "update_bugs",
+)
+
+_CLIENT = None
+
+
+def get_client():
+    """Get bugzilla service"""
+    global _CLIENT  # pylint: disable=global-statement
+    if not _CLIENT:
```

Comment:
Wouldn't have `functools.lru_cache()` have worked here?

Reply:
I think I avoided `lru_cache` at first because the Redhat Bugzilla client had side effects on instantiation, but now that we have our own client that doesn't do this and the Jira client also doesn't have any side effects on instantiation, `lru_cache` does seem nicer.